File: craft_project_invoicing/public/js/sales_invoice.js
frappe.ui.form.on('Sales Invoice', {
    // custom_invoice_percentage:function(frm){
    //     if (frm.doc.custom_invoice_percentage > 100) {
    //         frappe.throw(__("Invoice percentage cannot be greater than 100"));
    //     }

    //     if (frm.doc.sales_order && frm.doc.custom_invoice_percentage) {
    //         frappe.call({
    //             method: "frappe.get_list",
    //             args: {
    //                 doctype: "Sales Invoice",
    //                 filters: {
    //                     sales_order: frm.doc.sales_order,
    //                     docstatus: 1,
    //                     name: ["!=", frm.doc.name]  // Exclude current invoice
    //                 },
    //                 fields: ["name"]
    //             },
    //             callback: function (r) {
    //                 if (r.message && r.message.length > 0) {
    //                     let invoice_names = r.message.map(d => d.name);

    //                     frappe.call({
    //                         method: "craft_project_invoicing.events.sales_invoice.get_invoice_percentage_sum",
    //                         args: {
    //                             invoice_names: invoice_names
    //                         },
    //                         callback: function (res) {
    //                             let previous_total = flt(res.message || 0);
    //                             let new_total = previous_total + flt(frm.doc.custom_invoice_percentage);
    //                             if (new_total > 100) {
    //                                 frappe.throw(__("Total invoice percentage exceeds 100%. Currently: {0}%", [new_total]));
    //                             }
    //                         }
    //                     });
    //                 }
    //             }
    //         });
    //     }
    // },
	onload_post_render: function (frm) {
		if (frm.doc.sales_order && frm.doc.__islocal) {
			frappe.db.get_doc("Sales Order", frm.doc.sales_order)
				.then(doc => {
					if (doc.enable_project_invoicing) {
						let adv_amount = doc.base_net_total * (doc.advance_percentage / 100);
						let delivery_amount = doc.base_net_total * (doc.on_delivery_percentage / 100);
						let balance_delivery_amount = 0;
						if (doc.advance_billed && !doc.retention_billed) {
							balance_delivery_amount = delivery_amount - (doc.total_billed_amount - adv_amount);
						}
						let retention = doc.base_net_total * (doc.retention_percentage / 100);

						// Advance Billing
						if (!doc.advance_billed && doc.advance_percentage && doc.base_net_total) {
							frappe.db.get_value("Item", { item_name: "Advance" }, "name")
								.then(res => {
									if (!res.message || !res.message.name) {
										frappe.throw(__("Item with item name 'Advance' not found."));
									}
									return frappe.db.get_doc("Item", res.message.name);
								})
								.then(item_doc => {
									let income_account = '';
									$.each(item_doc.item_defaults, function (k, i) {
										if (i.company == frm.doc.company) {
											income_account = i.income_account;
										}
									});
									let rate = adv_amount;
									frm.clear_table("items");
									let row = frm.add_child("items", {
										item_code: item_doc.name,
										item_name: item_doc.item_name,
										rate: rate,
										qty: 1,
										conversion_factor: 1,
										uom: item_doc.stock_uom,
										description: item_doc.description,
										sales_order: doc.sales_order,
										income_account: income_account,
										project: frm.doc.project,
										cost_center: frm.doc.cost_center
									});
									frm.trigger('rate', row.doctype, row.name);
									frm.refresh_fields("items");
								});
							set_taxes(frm, frm.doc.sales_order);
						}
						// Retention Billing
						else if (doc.advance_billed && doc.on_delivery_billed && !doc.retention_billed && doc.retention_percentage && doc.base_net_total) {
							frappe.db.get_value("Item", { item_name: "Retention" }, "name")
								.then(res => {
									if (!res.message || !res.message.name) {
										frappe.throw(__("Item with item name 'Retention' not found."));
									}
									return frappe.db.get_doc("Item", res.message.name);
								})
								.then(item_doc => {
									let income_account = '';
									$.each(item_doc.item_defaults, function (k, i) {
										if (i.company == frm.doc.company) {
											income_account = i.income_account;
										}
									});
									let rate = doc.base_net_total * (doc.retention_percentage / 100);
									frm.clear_table("items");
									let row = frm.add_child("items", {
										item_code: item_doc.name,
										item_name: item_doc.item_name,
										rate: rate,
										qty: 1,
										conversion_factor: 1,
										uom: item_doc.stock_uom,
										description: item_doc.description,
										sales_order: doc.sales_order,
										income_account: income_account,
										project: frm.doc.project,
										cost_center: frm.doc.cost_center
									});
									frm.trigger('rate', row.doctype, row.name);
									frm.refresh_fields("items");
									if (!frm.doc.taxes_and_charges) {
										set_taxes(frm, frm.doc.sales_order);
									}
								});
						} else {
							frappe.db.get_value("Company", frm.doc.company, "project_invoicing_tax_template")
								.then(r => {
									if (r.message) {
										frm.set_value("taxes_and_charges", r.message.project_invoicing_tax_template);
										setTimeout(() => {
											$.each(frm.doc.taxes, function (k, t) {
												if (t.is_advance && !t.is_retention) {
													frappe.model.set_value(t.doctype, t.name, "rate", -(doc.advance_percentage));
												} else if (!t.is_advance && t.is_retention) {
													frappe.model.set_value(t.doctype, t.name, "rate", -(doc.retention_percentage));
												}
											});
										}, 1000);
									}
									else {
										frappe.throw({ message: __("Project invoicing template not found in company"), title: __("Message") })
									}
								});
						}
					}
				});
		}
	},

	// validate: function (frm) {
    //     if (frm.doc.items && frm.doc.items.length > 0) {
    //         let promises = [];
    
    //         $.each(frm.doc.items, function (index, row) {
    //             if (!row.sales_order) return;
    
    //             let p = new Promise(resolve => {
    //                 frappe.call({
    //                     method: "craft_project_invoicing.events.sales_invoice.get_so_detail",
    //                     args: {
    //                         "sales_order": row.sales_order,
    //                         "invoice_per": frm.doc.custom_invoice_percentage,
    //                         "doc": frm.doc
    //                     },
    //                     callback: function (r) {
    //                         if (r.message) {
    //                             if (r.message[row.so_detail]) {
    //                                 frappe.model.set_value(row.doctype, row.name, "custom_so_qty", r.message[row.so_detail].so_qty);
    //                                 if (r.message[row.so_detail].qty) {
    //                                     frappe.model.set_value(row.doctype, row.name, "qty", r.message[row.so_detail].qty);
    //                                 }
    //                             }
    //                             if (frm.doc.custom_invoice_percentage && !row.invoicing_percentage) {
    //                                 frappe.model.set_value(row.doctype, row.name, "invoicing_percentage", frm.doc.custom_invoice_percentage);
    //                             }
    //                         }
    //                         frappe.call({
    //                             method: "frappe.client.get_value",
    //                             args: {
    //                                 doctype: "Sales Order",
    //                                 filters: { name: row.sales_order },
    //                                 fieldname: ["advance_percentage", "retention_percentage"]
    //                             },
    //                             callback: function (r2) {
    //                                 if (r2.message) {
    //                                     let advance_percent = r2.message.advance_percentage || 0;
    //                                     let retention_percent = r2.message.retention_percentage || 0;
    //                                     let qty = row.qty || 0;
    //                                     let rate = row.rate || 0;
    //                                     let calculated_amount = qty * rate;
    
    //                                     frappe.model.set_value(row.doctype, row.name, "advance_percent", advance_percent);
    //                                     frappe.model.set_value(row.doctype, row.name, "retention_percent", retention_percent);
    
    //                                     let advance_amount = (calculated_amount * advance_percent) / 100;
    //                                     frappe.model.set_value(row.doctype, row.name, "advance_amount", advance_amount);
    //                                 }
    //                                 resolve();
    //                             }
    //                         });
    //                     }
    //                 });
    //             });
    //             promises.push(p);
    //         });
    
    //         Promise.all(promises).then(() => {
    //             frm.refresh_field("items");
    //         });
    //     }
    // }

    validate: async function (frm) {
        if (!frm.doc.items || frm.doc.items.length === 0) return;

        if (frm.doc.custom_invoice_percentage && frm.doc.custom_invoice_percentage > 100) {
            frappe.throw(__("Invoice percentage cannot be greater than 100"));
        }

        frm.doc.items.forEach(row => {
            if (row.invoicing_percentage && row.invoicing_percentage > 100) {
                frappe.throw(__("Row #{0}: Invoicing Percentage cannot be greater than 100%", [row.idx]));
            }
        });

        
        if (frm.doc.sales_order && frm.doc.custom_invoice_percentage) {
            const items_payload = (frm.doc.items || []).map(d => ({
                so_detail: d.so_detail || d.sales_order_item,
                item_code: d.item_code,
                invoicing_percentage: flt(d.invoicing_percentage || frm.doc.custom_invoice_percentage)
            }));

            const res = await frappe.call({
                method: "craft_project_invoicing.events.sales_invoice.validate_invoice_percentage_total",
                args: {
                    sales_order: frm.doc.sales_order,
                    current_invoice: frm.doc.name,
                    custom_invoice_percentage: frm.doc.custom_invoice_percentage,
                    items: items_payload
                }
            });

            const out = res.message || {};
            const total_invoice_percent = flt(out.total_invoice_percent);

            if (total_invoice_percent > 100) {
                frappe.throw(__("Total invoice percentage exceeds 100%. Currently: {0}%", [total_invoice_percent]));
            }

            if (out.items_exceeding && out.items_exceeding.length > 0) {
                const lines = out.items_exceeding.map((d, idx) =>
                    `Row #${idx + 1}: Invoicing percentage exceeds 100%.<br>` +
                    `Total : <b>${d.total}%</b> (Current: ${d.current}%, Previous: ${d.prev}%)`
                );
                frappe.throw(lines.join("<br><br>"));
            }



        }

        let promises = frm.doc.items.map(async (row) => {
            if (row.sales_order) {
                let so_detail_resp = await frappe.call({
                    method: "craft_project_invoicing.events.sales_invoice.get_so_detail",
                    args: {
                        sales_order: row.sales_order,
                        invoice_per: frm.doc.custom_invoice_percentage,
                        doc: frm.doc
                    }
                });
    
                if (so_detail_resp.message && so_detail_resp.message[row.so_detail]) {
                    frappe.model.set_value(row.doctype, row.name, "custom_so_qty", so_detail_resp.message[row.so_detail].so_qty);
                    if (so_detail_resp.message[row.so_detail].qty) {
                        frappe.model.set_value(row.doctype, row.name, "qty", so_detail_resp.message[row.so_detail].qty);
                    }
                }
    
                if (frm.doc.custom_invoice_percentage && !row.invoicing_percentage) {
                    frappe.model.set_value(row.doctype, row.name, "invoicing_percentage", frm.doc.custom_invoice_percentage);
                }
    
                let get_value_resp = await frappe.call({
                    method: "frappe.client.get_value",
                    args: {
                        doctype: "Sales Order",
                        filters: { name: row.sales_order },
                        fieldname: ["advance_percentage", "retention_percentage"]
                    }
                });
    
                if (get_value_resp.message) {
                    let advance_percent = get_value_resp.message.advance_percentage || 0;
                    let retention_percent = get_value_resp.message.retention_percentage || 0;
                    let calculated_amount = (row.qty || 0) * (row.rate || 0);
    
                    frappe.model.set_value(row.doctype, row.name, "advance_percent", advance_percent);
                    frappe.model.set_value(row.doctype, row.name, "retention_percent", retention_percent);
    
                    frappe.model.set_value(row.doctype, row.name, "advance_amount", (calculated_amount * advance_percent) / 100);
                    frappe.model.set_value(row.doctype, row.name, "retention_amount", (calculated_amount * retention_percent) / 100);
                }
            }
        });
    
        await Promise.all(promises);
    
        let total_advance_amount = frm.doc.items.reduce((sum, row) => {
            let calculated_amount = (row.qty || 0) * (row.rate || 0);
            let advance_amount = (calculated_amount * (row.advance_percent || 0)) / 100;
            return sum + advance_amount;
        }, 0);
    
        let total_retention_amount = frm.doc.items.reduce((sum, row) => {
            let calculated_amount = (row.qty || 0) * (row.rate || 0);
            let retention_amount = (calculated_amount * (row.retention_percent || 0)) / 100;
            return sum + retention_amount;
        }, 0);
    
        if (frm.doc.taxes && frm.doc.taxes.length > 0) {
            frm.doc.taxes.forEach(tax => {
                if (tax.is_advance == 1) {
                    frappe.model.set_value(tax.doctype, tax.name, "tax_amount", -total_advance_amount);
                }
                if (tax.is_retention == 1) {
                    frappe.model.set_value(tax.doctype, tax.name, "tax_amount", -total_retention_amount);
                }
            });
            frm.refresh_field("taxes");
        }
    
        frm.refresh_field("items");
    }
});

var set_taxes = function(frm, sales_order) {
	frappe.db.get_value("Sales Order", sales_order, "taxes_and_charges")
	.then(r => {
		if (r.message.taxes_and_charges) {
			setTimeout(() => {
				frm.set_value("taxes_and_charges", r.message.taxes_and_charges);
				frm.trigger("taxes_and_charges", frm.doc.doctype, frm.doc.name);
			}, 1000);
		}
	})
};

File: craft_project_invoicing/events/sales_invoice.py
import frappe
from frappe.utils import flt


def on_submit(doc, method):
	if doc.items and doc.sales_order:
		so_doc = frappe.get_doc("Sales Order", doc.sales_order)
		if so_doc.enable_project_invoicing:
			retention_account = frappe.db.get_value(
				"Item Default", {"parent": "Retention", "company": doc.company}, "income_account")
			adv_account = frappe.db.get_value(
				"Item Default", {"parent": "Advance", "company": doc.company}, "income_account")

			# Post reverse journal entry for advance invoice
			if doc.items and len(doc.items) == 1 and doc.items[0].item_code == "Advance":
				if not adv_account:
					frappe.throw(
						title="Advance Account Not Found",
						msg="Deafult income account not found in item <b>Advance</b>"
					)
				adv_amount = doc.items[0].get("amount")
				if not doc.items[0].get("income_account") == adv_account:
					ge_doc = frappe.get_doc({
						"doctype": "Journal Entry",
						"voucher_type": "Journal Entry",
						"company": doc.company,
						"posting_date": doc.posting_date,
						"sales_order_reference": doc.sales_order
					})
					ge_doc.append("accounts", {
						"account": doc.items[0].get("income_account"),
						"debit_in_account_currency": adv_amount,
						"credit_in_account_currency": 0,
						"project": doc.project,
						"cost_center": doc.items[0].get("cost_center")
					})
					ge_doc.append("accounts", {
						"account": adv_account,
						"debit_in_account_currency": 0,
						"credit_in_account_currency": adv_amount,
						"project": doc.project,
						"cost_center": doc.items[0].get("cost_center")
					})
					ge_doc.insert(ignore_permissions=True,
								  ignore_mandatory=True)
					ge_doc.submit()
					doc.db_set("ref_journal_entry", ge_doc.name)

				so_doc.db_set({
					"advance_billed": 1,
					"advance_ref_doc": doc.name,
					"advance_billed_amount": adv_amount,
					# "remaining_advance": adv_amount
				})

			# Post reverse journal entry for on Retention invoice
			elif doc.items and len(doc.items) == 1 and doc.items[0].item_code == "Retention":
				if not retention_account:
					frappe.throw(
						title="Retention Account Not Found",
						msg="Deafult income account not found in item <b>Retention</b>"
					)
				ret_amount = doc.items[0].get("amount")
				if not doc.items[0].get("income_account") == retention_account:
					ge_doc = frappe.get_doc({
						"doctype": "Journal Entry",
						"voucher_type": "Journal Entry",
						"company": doc.company,
						"posting_date": doc.posting_date,
						"sales_order_reference": doc.sales_order
					})
					ge_doc.append("accounts", {
						"account": doc.items[0].get("income_account"),
						"debit_in_account_currency": ret_amount,
						"credit_in_account_currency": 0,
						"project": doc.project,
						"cost_center": doc.items[0].get("cost_center")
					})
					ge_doc.append("accounts", {
						"account": retention_account,
						"debit_in_account_currency": 0,
						"credit_in_account_currency": ret_amount,
						"project": doc.project,
						"cost_center": doc.items[0].get("cost_center")
					})
					ge_doc.insert(ignore_permissions=True,
								  ignore_mandatory=True)
					ge_doc.submit()
					doc.db_set("ref_journal_entry", ge_doc.name)

				so_doc.db_set({
					"retention_billed": 1,
					"retention_ref_doc": doc.name,
					"retention_billed_amount": ret_amount,
					# "remaining_retention": 0
				})

			# Post reverse journal entry for on multiple delivery invoices
			elif doc.items and len(doc.items) > 0:
				actual_delivery_amount = so_doc.base_net_total * \
					so_doc.on_delivery_percentage / 100
				so_doc.db_set("delivery_ref_doc", doc.name if not so_doc.delivery_ref_doc else (
					str(so_doc.delivery_ref_doc) + ", " + str(doc.name)))
				total = 0
				consumed_advance = 0
				# remaining_advance = 0
				consumed_retention = 0
				# remaining_retention = 0

				for t in doc.taxes:
					if t.is_advance and t.tax_amount:
						consumed_advance = so_doc.consumed_advance + \
							abs(t.tax_amount)
						# remaining_advance = so_doc.advance_billed_amount - (consumed_advance if consumed_advance else 0)
					elif t.is_retention and t.tax_amount:
						consumed_retention = so_doc.consumed_retention + \
							abs(t.tax_amount)
						# remaining_retention = (so_doc.base_net_total * (so_doc.retention_percentage/100)) - (consumed_retention if consumed_retention else 0)

					if t.is_advance or t.is_retention:
						total = total + t.tax_amount
				if actual_delivery_amount <= (float(so_doc.delivery_billed_amount if so_doc.delivery_billed_amount else 0) + doc.total + total):
					so_doc.db_set("on_delivery_billed", 1)
				so_doc.db_set({
					"delivery_billed_amount": so_doc.delivery_billed_amount + (doc.total + total),
					"consumed_advance": consumed_advance,
					"consumed_retention": consumed_retention,
					# "remaining_advance": remaining_advance,
					# "remaining_retention": remaining_retention
				})


# def validate(doc, method):
#     if doc.items and doc.sales_order:
#         for item in doc.items:
#             previous_data = frappe.db.sql("""
#                 SELECT SUM(sii.amount), SUM(sii.invoicing_percentage)
#                 FROM `tabSales Invoice Item` sii
#                 JOIN `tabSales Invoice` si ON sii.parent = si.name
#                 WHERE sii.item_code = %s 
#                 AND si.sales_order = %s 
#                 AND si.docstatus = 1
#                 AND si.name != %s
#             """, (item.item_code, doc.sales_order, doc.name))

#             previous_amount = previous_data[0][0] if previous_data and previous_data[0][0] else 0
#             previous_percentage = previous_data[0][1] if previous_data and previous_data[0][1] else 0

#             item.previous_amount = previous_amount
#             item.previous_percentage = previous_percentage
#             item.cumulative_amount = item.amount + previous_amount
#             item.cumulative_percentage = item.invoicing_percentage + previous_percentage



#     if doc.taxes and doc.sales_order:
#         for tax in doc.taxes:
#             condition = ""
#             params = [doc.sales_order, doc.name]

#             if tax.is_advance:
#                 condition = "AND stc.is_advance = 1"
#             elif tax.is_retention:
#                 condition = "AND stc.is_retention = 1"

#             previous_tax_amount = frappe.db.sql(f"""
#                 SELECT SUM(tax_amount) 
#                 FROM `tabSales Taxes and Charges` stc
#                 JOIN `tabSales Invoice` si ON stc.parent = si.name
#                 WHERE si.sales_order = %s 
#                 AND si.docstatus = 1
#                 AND si.name != %s
#                 {condition}
#             """, tuple(params))

#             tax.previous_amount = previous_tax_amount[0][0] if previous_tax_amount and previous_tax_amount[0][0] else 0
#             tax.cumulative_amount = tax.tax_amount + tax.previous_amount



def on_cancel(doc, method):
	invoice_type = {
		"Advance": "advance_billed",
		"Retention": "retention_billed"
	}
	invoice_references = {
		"Advance": "advance_ref_doc",
		"Retention": "retention_ref_doc"
	}
	billing_amounts = {
		"Advance": "advance_billed_amount",
		"Retention": "retention_billed_amount",
		"Delivery": "delivery_billed_amount"
	}
	if doc.ref_journal_entry:
		jv_doc_list = doc.ref_journal_entry.split(", ")
		for j in jv_doc_list:
			jv_doc = frappe.get_doc("Journal Entry", j)
			jv_doc.cancel()
	if doc.sales_order:
		so_doc = frappe.get_doc("Sales Order", doc.sales_order)
		if so_doc.enable_project_invoicing:
			delivery_ref_docs = None
			if doc.items and len(doc.items) == 1 and doc.items[0].item_code in ["Advance", "Retention"]:
				frappe.db.set_value("Sales Order", doc.sales_order, {
					invoice_type.get(doc.items[0].item_code): 0,
					invoice_references.get(doc.items[0].item_code): "",
					billing_amounts.get(doc.items[0].item_code): 0
				})
				if doc.items[0].item_code == "Retention":
					frappe.db.set_value("Sales Order", doc.sales_order, {
						"remaining_retention": float(so_doc.remaining_retention) - (doc.items[0].get("amount") if doc.items[0].get("amount") else 0)
					})
			elif doc.items and len(doc.items) > 0 and not doc.items[0].item_code in ["Advance", "Retention"]:
				frappe.db.set_value(
					"Sales Order", doc.sales_order, "on_delivery_billed", 0)
				delivery_ref_docs = frappe.db.get_value(
					"Sales Order", doc.sales_order, "delivery_ref_doc")

			if delivery_ref_docs:
				delivery_ref_doc_list = delivery_ref_docs.split(", ")
				delivery_ref_doc_list.remove(doc.name)
				frappe.db.set_value("Sales Order", doc.sales_order,
									"delivery_ref_doc", ", ".join(delivery_ref_doc_list))

				delivery_billed_amount = frappe.db.get_value(
					"Sales Order", doc.sales_order, "delivery_billed_amount")
				if delivery_billed_amount:
					total = 0
					consumed_advance = 0
					# remaining_advance = 0
					consumed_retention = 0
					# remaining_retention = 0
					for t in doc.taxes:
						if t.is_advance and t.tax_amount:
							consumed_advance = so_doc.consumed_advance - \
								abs(t.tax_amount)
							# remaining_advance = so_doc.advance_billed_amount - (consumed_advance if consumed_advance else 0)
						elif t.is_retention and t.tax_amount:
							consumed_retention = so_doc.consumed_retention - \
								abs(t.tax_amount)
							# remaining_retention = so_doc.retention_billed_amount - (consumed_retention if consumed_retention else 0)
						if t.is_advance or t.is_retention:
							total = total + t.tax_amount
					frappe.db.set_value("Sales Order", doc.sales_order, {
						billing_amounts.get("Delivery"): delivery_billed_amount - (doc.total + total),
						"consumed_advance": consumed_advance,
						"consumed_retention": consumed_retention,
						# "remaining_advance": remaining_advance,
						# "remaining_retention": remaining_retention
					})


import json

@frappe.whitelist()
def get_so_detail(sales_order, invoice_per=None, doc=None):
    if not sales_order:
        return None

    if doc:
        if isinstance(doc, str):
            doc = json.loads(doc)
        doc = frappe.get_doc(doc)

    invoice_per = flt(invoice_per) if invoice_per else None
    so_doc = frappe.get_doc("Sales Order", sales_order)

    if not so_doc.items:
        frappe.throw("Sales Order has no items or does not exist.")

    so_details = {}
    for item in so_doc.items:
        so_details[item.name] = {
            "so_detail": item.name,
            "so_qty": item.qty,
            "qty": item.qty
        }

    for item in doc.items:
        so_detail_key = item.get("so_detail")

        if so_detail_key and so_detail_key in so_details:
            invoice_percentage = flt(item.get("invoicing_percentage")) or invoice_per
            if invoice_percentage:
                so_details[so_detail_key]["qty"] = so_details[so_detail_key]["so_qty"] * (invoice_percentage / 100)

    return so_details if so_details else None


@frappe.whitelist()
def get_invoice_items(invoice_names, item_code, sales_order):
    invoice_names = frappe.parse_json(invoice_names)
    return frappe.get_all("Sales Invoice Item", 
                          filters={"parent": ["in", invoice_names], "item_code": item_code, "sales_order": sales_order},
                          fields=["amount", "invoicing_percentage","advance_amount","balance_amount_ad","retention_amount","balance_amount_ret"])
